File: aufgaben/a1/src/ram/program_memory.c
#include <stdio.h>
#include <string.h>
#include "../cpu/instructions.h"
#include "../stack/stack.h"
#include "program_memory.h"

int pc = 0;
uint32_t program_memory[MAXITEMS];

void print_memory(void) {
    for (int i = 0; i < pc; i++) {
        Instruction instruction = decode_instruction(program_memory[i]);
        Opcode opcode = instruction.opcode;
        int immediate = instruction.immediate;
        switch (opcode) {
            case halt:
                printf("%03d:\thalt\n", i);
                break;
            case pushc:
                printf("%03d:\tpushc   %d\n", i, immediate);
                break;
            case add:
                printf("%03d:\tadd\n", i);
                break;
            case sub:
                printf("%03d:\tsub\n", i);
                break;
            case mul:
                printf("%03d:\tmul\n", i);
                break;
            case divide:
                printf("%03d:\tdiv\n", i);
                break;
            case mod:
                printf("%03d:\tmod\n", i);
                break;
            case rdint:
                printf("%03d:\trdint\n", i);
                break;
            case wrint:
                printf("%03d:\twrint\n", i);
                break;
            case rdchr:
                printf("%03d:\trdchr\n", i);
                break;
            case wrchr:
                printf("%03d:\twrchr\n", i);
                break;
            default:
                break;
        }
    }
}

void register_instruction(Opcode opcode, int immediate) {
    uint32_t instruction = encode_instruction(opcode, immediate);
    program_memory[pc] = instruction;
    pc++;
}

File: src/memory/program_memory.rs
use crate::{Bytecode, Immediate, Instruction, Opcode, Opcode::*};

#[derive(Debug, Eq, PartialEq, Clone)]
pub struct ProgramMemory<U> {
    pub pc: u32,
    pub memory: Vec<U>,
}

impl Default for ProgramMemory<Bytecode>
where
    Bytecode: std::fmt::Debug + std::fmt::Display,
{
    fn default() -> Self {
        Self::new()
    }
}

impl ProgramMemory<Bytecode> {
    pub fn new() -> Self {
        ProgramMemory { pc: 0, memory: vec![] }
    }
    pub fn register_instruction(&mut self, opcode: Opcode, immediate: Immediate) {
        let instruction: Bytecode = Instruction::encode_instruction(opcode, immediate);
        self.memory.push(instruction);
        self.pc += 1;
    }
    pub fn print(&self) {
        for i in 0..self.pc {
            let instruction: Instruction = Instruction::decode_instruction(self.memory[i as usize]);
            let immediate = instruction.immediate;
            match instruction.opcode {
                Halt => println!("{i:03}:\thalt"),
                Pushc => println!("{i:03}:\tpushc\t{}", immediate),
                Add => println!("{i:03}:\tadd"),
                Sub => println!("{i:03}:\tsub"),
                Mul => println!("{i:03}:\tmul"),
                Div => println!("{i:03}:\tdiv"),
                Mod => println!("{i:03}:\tmod"),
                Rdint => println!("{i:03}:\trdint"),
                Wrint => println!("{i:03}:\twrint"),
                Rdchr => println!("{i:03}:\trdchr"),
                Wrchr => println!("{i:03}:\twrchr"),
                Pushg => println!("{i:03}:\tpushg\t{}", immediate),
                Popg => println!("{i:03}:\tpopg\t{}", immediate),
                Asf => println!("{i:03}:\tasf\t{}", immediate),
                Rsf => println!("{i:03}:\trsf"),
                Pushl => println!("{i:03}:\tpushl\t{}", immediate),
                Popl => println!("{i:03}:\tpopl\t{}", immediate),
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::{Opcode::Pushc, ProgramMemory};
    #[test]
    fn test_program_memory() {
        let program_memory = ProgramMemory::default();
        assert_eq!(program_memory.pc, 0);
        assert_eq!(program_memory.memory.len(), 0);
    }
    #[test]
    fn test_register_instruction() {
        let mut program_memory = ProgramMemory::default();
        program_memory.register_instruction(Pushc, 1);
        assert_eq!(program_memory.pc, 1);
        assert_eq!(program_memory.memory[0], 0x01000001);
        program_memory.register_instruction(Pushc, 2);
        assert_eq!(program_memory.pc, 2);
        assert_eq!(program_memory.memory[1], 0x01000002);
    }
}
